File: frontend/src/store/projetos.ts
import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { Projeto } from '../types';
import { projetosService, ProjetoCreateParams } from '../services/projetos';

interface ProjetosState {
  projetos: Projeto[];
  projetoAtivo: Projeto | null;
  loading: boolean;
  erro: string | null;

  listar: () => Promise<void>;
  criar: (dados: ProjetoCreateParams) => Promise<Projeto>;
  selecionarProjeto: (id: string) => Promise<void>;
  atualizarProjeto: (id: string, dados: Partial<ProjetoCreateParams>) => Promise<Projeto>;
  atualizarFases: (fases: { id: string; percentual: number }[]) => Promise<void>;
  atualizarFaseUnica: (faseId: string, percentual: number) => Promise<void>;
  limparErro: () => void;
}

export const useProjetosStore = create<ProjetosState>()(
  persist(
    (set, get) => ({
      projetos: [],
      projetoAtivo: null,
      loading: false,
      erro: null,

      listar: async () => {
        set({ loading: true, erro: null });
        try {
          const projetos = await projetosService.listar();
          set({ projetos, loading: false });
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao carregar lista de projetos.';
          set({ erro: mensagem, loading: false });
        }
      },

      criar: async (dados) => {
        set({ loading: true, erro: null });
        try {
          const novoProjeto = await projetosService.criar(dados);
          set((state) => ({
            projetos: [novoProjeto, ...state.projetos],
            projetoAtivo: novoProjeto,
            loading: false
          }));
          return novoProjeto;
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao criar projeto.';
          set({ erro: mensagem, loading: false });
          throw error;
        }
      },

      selecionarProjeto: async (id) => {
        set({ loading: true, erro: null });
        try {
          const projeto = await projetosService.obter(id);
          set({ projetoAtivo: projeto, loading: false });
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao carregar detalhes do projeto.';
          set({ erro: mensagem, loading: false });
          throw error;
        }
      },

      atualizarProjeto: async (id, dados) => {
        set({ loading: true, erro: null });
        try {
          const projetoAtualizado = await projetosService.atualizar(id, dados);
          set((state) => ({
            projetos: state.projetos.map(p => p.id === id ? { ...p, ...projetoAtualizado } : p),
            projetoAtivo: state.projetoAtivo?.id === id ? projetoAtualizado : state.projetoAtivo,
            loading: false
          }));
          return projetoAtualizado;
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao atualizar configurações do projeto.';
          set({ erro: mensagem, loading: false });
          throw error;
        }
      },

      atualizarFases: async (fases) => {
        const { projetoAtivo } = get();
        if (!projetoAtivo) return;

        set({ loading: true, erro: null });
        try {
          await projetosService.atualizarFases(projetoAtivo.id, fases);
          // Recarrega o projeto ativo para atualizar as fases e o progresso
          const projetoAtualizado = await projetosService.obter(projetoAtivo.id);
          set({ projetoAtivo: projetoAtualizado, loading: false });
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao atualizar fases.';
          set({ erro: mensagem, loading: false });
          throw error;
        }
      },

      atualizarFaseUnica: async (faseId, percentual) => {
        const { projetoAtivo } = get();
        if (!projetoAtivo) return;

        set({ loading: true, erro: null });
        try {
          await projetosService.atualizarFaseUnica(projetoAtivo.id, faseId, percentual);
          const projetoAtualizado = await projetosService.obter(projetoAtivo.id);
          set({ projetoAtivo: projetoAtualizado, loading: false });
        } catch (error: any) {
          const mensagem = error.response?.data?.detail || 'Erro ao atualizar fase.';
          set({ erro: mensagem, loading: false });
          throw error;
        }
      },

      limparErro: () => set({ erro: null }),
    }),
    {
      name: 'appcontrol-projetos-storage',
      partialize: (state) => ({ projetoAtivo: state.projetoAtivo }),
    }
  )
);

File: backend/app/routers/projetos.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from app.models.schemas import ProjetoCreate, ProjetoUpdate, FaseUpdate
from app.services.projetos_service import ProjetosService
from app.services.auth_service import AuthService
from app.routers.auth import get_token_from_header
from typing import List, Dict, Any

router = APIRouter()

class FasePercentualUpdate(BaseModel):
    percentual: int = Field(ge=0, le=100, description="Percentual de conclusão da fase (0 a 100)")

async def get_current_usuario_id(token: str = Depends(get_token_from_header)) -> str:
    """
    Dependência para validar o token JWT e extrair o ID do usuário autenticado.
    """
    usuario = AuthService.get_current_user(token)
    return str(usuario.id)

@router.get("", summary="Listar Projetos")
async def listar_projetos(usuario_id: str = Depends(get_current_usuario_id)) -> List[Dict[str, Any]]:
    """
    Retorna a lista de projetos do usuário autenticado, enriquecida com progresso e contagem de sessões.
    """
    return ProjetosService.listar_projetos(usuario_id)

@router.post("", summary="Criar Projeto", status_code=status.HTTP_201_CREATED)
async def criar_projeto(dados: ProjetoCreate, usuario_id: str = Depends(get_current_usuario_id)) -> Dict[str, Any]:
    """
    Cria um novo projeto e gera automaticamente as 6 fases padrão do roadmap.
    """
    return ProjetosService.criar_projeto(usuario_id, dados)

@router.get("/{id}", summary="Obter Detalhes do Projeto")
async def obter_projeto(id: str, usuario_id: str = Depends(get_current_usuario_id)) -> Dict[str, Any]:
    """
    Retorna os dados completos de um projeto (incluindo fases e sessões).
    """
    return ProjetosService.obter_projeto(usuario_id, id)

@router.put("/{id}", summary="Atualizar Configurações do Projeto")
async def atualizar_projeto(id: str, dados: ProjetoUpdate, usuario_id: str = Depends(get_current_usuario_id)) -> Dict[str, Any]:
    """
    Atualiza as configurações (stacks, LLM, arquivos críticos, regras especiais) de um projeto.
    """
    return ProjetosService.atualizar_projeto(usuario_id, id, dados)

@router.get("/{id}/fases", summary="Listar Fases do Projeto")
async def listar_fases(id: str, usuario_id: str = Depends(get_current_usuario_id)) -> List[Dict[str, Any]]:
    """
    Retorna as fases de roadmap de um projeto.
    """
    return ProjetosService.listar_fases(usuario_id, id)

@router.put("/{id}/fases", summary="Atualizar Fases em Lote")
async def atualizar_fases(id: str, dados: FaseUpdate, usuario_id: str = Depends(get_current_usuario_id)) -> Dict[str, Any]:
    """
    Atualiza o percentual de múltiplas fases simultaneamente.
    """
    return ProjetosService.atualizar_fases(usuario_id, id, dados)

@router.put("/{id}/fases/{fase_id}", summary="Atualizar Fase Única")
async def atualizar_fase(id: str, fase_id: str, dados: FasePercentualUpdate, usuario_id: str = Depends(get_current_usuario_id)) -> Dict[str, Any]:
    """
    Atualiza o percentual de uma única fase específica.
    """
    return ProjetosService.atualizar_fase_unica(usuario_id, id, fase_id, dados.percentual)
